provider/spark.go

```diff
@@ -944,40 +972,91 @@ func (s *S3Materialization) NumRows() (int64, error) {
 }
 
 func (s *S3Materialization) IterateSegment(begin, end int64) (FeatureIterator, error) {
-	stream, err := s.store.RowStreamFromSelectQuery(s.Key, fmt.Sprintf("SELECT * FROM S3Object WHERE row_number > %d AND row_number <= %d", begin, end))
+	stream, err := s.store.ResourceStreamConv(s.Key)
 	if err != nil {
 		return nil, err
 	}
+	for i := int64(0); i < begin; i++ {
+		_ = <-stream
+	}
 	return &S3FeatureIterator{stream: stream, maxIdx: (end - begin)}, nil
 }
 
 type S3FeatureIterator struct {
-	stream chan []byte
+	stream chan interface{}
 	cur    ResourceRecord
 	err    error
 	curIdx int64
 	maxIdx int64
 }
 
-// expected format is "<entity(string)>,<value(interface{})>,<timestamp(int64)>"
-func featureCSVToResource(csv string) (ResourceRecord, error) {
-	values := strings.Split(csv, ",")
-	entity := string(values[0])
-	value := reflect.ValueOf(values[1]).Interface()
-	timeStampMilli, err := strconv.Atoi(values[2])
-	if err != nil {
-		return ResourceRecord{}, err
+func convIndirectToVal(value interface{}) interface{} {
+
+	if value == nil {
+		return nil
+	}
+	switch v := value.(type) {
+	case *string:
+		if v == nil {
+			return nil
+		}
+		return *v
+	case *int:
+		return *v
+	case *int32:
+		if v == nil {
+			return nil
+		}
+		return int(*v)
+	case *int64:
+		return *v
+	case *float32:
+		return *v
+	case *float64:
+		return *v
+	case *bool:
+		return *v
+	}
+	return nil
+}
+
+func featureStructToResource(row interface{}) (ResourceRecord, error) {
+	rowVal := reflect.ValueOf(row)
+	if rowVal.NumField() < 3 {
+		return ResourceRecord{}, fmt.Errorf("not enough fields in feature struct")
+	}
+	entity := *rowVal.Field(0).Interface().(*string)
+	value := convIndirectToVal(rowVal.Field(1).Interface())
+	//convert whatever time format is to time.Time
+	timestampValue := rowVal.Field(2).Interface()
+	var ts time.Time
+	switch v := timestampValue.(type) {
+	case *string:
+		//need to get non-converted value
+		layout := "2006-01-02T15:04:05.000Z"
```

Comment:
Whats this magic timestamp do?

Reply:
replaced it with the built in time template for interpreting timestamp string